TCF4.py

```diff
@@ -0,0 +1,121 @@
+from Bio import SeqIO
+from Bio.SeqUtils import GC
+
+def filter_fastq(input_path: str, quality_threshold: int, output_filename="final_filtered.fastq",  gc_bounds=(40, 60), length_bounds=(50, 350)):
+    filename = input_path
+    records = SeqIO.parse(filename, "fastq")
+    ###quality filter
+    good_reads = (rec for rec in records if min(rec.letter_annotations["phred_quality"]) >= quality_threshold)
+    result_quality = SeqIO.write(good_reads, "good_quality.fastq", "fastq")
+    result_quality_GC = SeqIO.parse("good_quality.fastq", "fastq")
+    ###GC content filter
+    min_gc_content = gc_bounds[0]
+    max_gc_content = gc_bounds[1]
+    GC_quality_filt = []
+    
+    for sequence in result_quality_GC:
+        if min_gc_content <= GC(sequence.seq) <= max_gc_content:
+            GC_quality_filt.append(sequence)
+            
+    result_quality = SeqIO.write(GC_quality_filt, "good_quality_GC.fastq", "fastq")
+    result_quality_GC_length = SeqIO.parse("good_quality_GC.fastq", "fastq")
+    
+    ##length filter
+    filtered_GC_quality_length = []
+    
+    for sequence in result_quality_GC_length:
+        if len(sequence.seq) >= length_bounds[0] and len(sequence.seq) <= length_bounds[1]:
+            filtered_GC_quality_length.append(sequence)
+            
+    result_quality = SeqIO.write(filtered_GC_quality_length, output_filename, "fastq")
+    
+    print(result_quality)
+
+#filter_fastq("example_fastq.fastq", 15)
+
+
+from abc import ABC, abstractmethod
+
+class InvalidInputError(ValueError):
+    pass
+
+class BiologicalSequence(ABC, str):
+    @abstractmethod
+    def __init__(self, seq):
+        self.seq = seq
```

Comment:
Насколько я успел посмотреть, в абстрактных классах не используют связку `__init__` и `@abstractmethod`.

Обычно делают так: в родительском, абстрактном классе объявляют переменную класса [`__slots__`](https://docs.python.org/3/reference/datamodel.html#slots). Это такой способ зарезервировать память под набор атрибутов. Затем в `__init__`  класса-наследника их определяют.

```suggestion
__slots__ = ('seq',)
```

Если вдруг станет интересно, что можно посмотреть:

1.   [Исходный код](https://github.com/python/cpython/blob/main/Lib/_collections_abc.py#L1006) абстрактного класса для последовательностей в Python.
2.  Особенности использования `__slots__` детально разобраны [тут](https://stackoverflow.com/a/28059785).

Reply:
upd: Я всё-таки [встретил](https://github.com/scikit-learn/scikit-learn/blob/2a2643f583812c9095f815564b1b346279c00a72/sklearn/ensemble/_forest.py#L740) связку `__init__` и `@abstractmethod`.